exchanges/kline/technical_analysis.go

```diff
@@ -0,0 +1,309 @@
+	errInvalidDataSetLengths      = errors.New("invalid data set lengths")
+)
+
+// OHLC is a connector for technical analysis usage
+type OHLC struct {
+	Open   []float64
+	High   []float64
+	Low    []float64
+	Close  []float64
+	Volume []float64
+}
+
+// GetOHLC returns the entire subset of candles as a friendly type for gct
+// technical analysis usage.
+func (k *Item) GetOHLC() *OHLC {
+	ohlc := &OHLC{
+		Open:   make([]float64, len(k.Candles)),
+		High:   make([]float64, len(k.Candles)),
+		Low:    make([]float64, len(k.Candles)),
+		Close:  make([]float64, len(k.Candles)),
+		Volume: make([]float64, len(k.Candles)),
+	}
+	for x := range k.Candles {
+		ohlc.Open[x] = k.Candles[x].Open
+		ohlc.High[x] = k.Candles[x].High
+		ohlc.Low[x] = k.Candles[x].Low
+		ohlc.Close[x] = k.Candles[x].Close
+		ohlc.Volume[x] = k.Candles[x].Volume
+	}
+	return ohlc
+}
+
+// GetAverageTrueRange returns the Average True Range for the given period.
+func (k *Item) GetAverageTrueRange(period int) ([]float64, error) {
+	return k.GetOHLC().GetAverageTrueRange(period)
+}
+
+// GetAverageTrueRange returns the Average True Range for the given period.
+func (o *OHLC) GetAverageTrueRange(period int) ([]float64, error) {
+	if o == nil {
+		return nil, fmt.Errorf("get average true range %w", errNilOHLC)
+	}
+	if period <= 0 {
+		return nil, fmt.Errorf("get average true range %w", errInvalidPeriod)
+	}
+	if len(o.High) == 0 {
+		return nil, fmt.Errorf("get average true range high %w", errNoData)
+	}
+	if len(o.Low) == 0 {
+		return nil, fmt.Errorf("get average true range low %w", errNoData)
+	}
+	if len(o.Close) == 0 {
+		return nil, fmt.Errorf("get average true range close %w", errNoData)
+	}
+	return indicators.ATR(o.High, o.Low, o.Close, period), nil
+}
+
+// GetBollingerBands returns Bollinger Bands for the given period.
+func (k *Item) GetBollingerBands(period int, nbDevUp, nbDevDown float64, m indicators.MaType) (*Bollinger, error) {
+	return k.GetOHLC().GetBollingerBands(period, nbDevUp, nbDevDown, m)
+}
+
+// Bollinger defines a return type for the bollinger bands
+type Bollinger struct {
+	Upper  []float64
+	Middle []float64
+	Lower  []float64
+}
+
+// GetBollingerBands returns Bollinger Bands for the given period.
+func (o *OHLC) GetBollingerBands(period int, nbDevUp, nbDevDown float64, m indicators.MaType) (*Bollinger, error) {
+	if o == nil {
+		return nil, fmt.Errorf("get bollinger bands %w", errNilOHLC)
+	}
+	if period <= 0 {
+		return nil, fmt.Errorf("get bollinger bands %w", errInvalidPeriod)
+	}
+	if nbDevUp <= 0 {
+		return nil, fmt.Errorf("get bollinger bands %w upper limit", errInvalidDeviationMultiplier)
+	}
+	if nbDevDown <= 0 {
+		return nil, fmt.Errorf("get bollinger bands %w lower limit", errInvalidDeviationMultiplier)
+	}
+	if len(o.Close) == 0 {
+		return nil, fmt.Errorf("get bollinger bands close %w", errNoData)
+	}
+	if period > len(o.Close) { // TODO: Investigate the panic when this protection is removed.
```

Comment:
The reason it panics is that there is code that hopes that the period is not > len and then uses `period` as an index 👀 
See
```go
func variance(inReal []float64, inTimePeriod int) []float64 {
	outReal := make([]float64, len(inReal))

	nbInitialElementNeeded := inTimePeriod - 1
	startIdx := nbInitialElementNeeded
	periodTotal1 := 0.0
	periodTotal2 := 0.0
	trailingIdx := startIdx - nbInitialElementNeeded
	i := trailingIdx
	if inTimePeriod > 1 {
                 // This bad boy here
		for i < startIdx {
                       // will ensure we lookup a slice which may not exist
			tempReal := inReal[i]
                        ...
		}
	}
```
So there's one of potentially a few possible problems with not checking the period in the indicators package 
# 👀 😄 😭   
I was expecting it to try and divide by zero...

Reply:
😭